warehouse/models/mart/gtfs/fct_monthly_shapes.sql

```diff
@@ -0,0 +1,70 @@
+{{ config(materialized='table') }}
+
+WITH dim_gtfs_datasets AS (
+    SELECT *
+    FROM {{ ref('dim_gtfs_datasets') }}
+),
+
+fct_scheduled_trips AS (
+    SELECT *
+    FROM {{ ref('fct_scheduled_trips') }}
+    WHERE service_date >= '2023-07-29' AND service_date <= '2023-07-31'
+),
+
+dim_shapes_arrays AS (
+
+    SELECT
+        key,
+        base64_url,
+        shape_id,
+        pt_array
+
+    FROM {{ ref('dim_shapes_arrays') }}
+    WHERE {{ incremental_where(default_start_var='GTFS_SCHEDULE_START',
+                               this_dt_column='_feed_valid_from',
+                               filter_dt_column='_feed_valid_from',
+                               dev_lookback_days = None)
+           }}
+),
+
+trips_by_shape AS (
+
+    SELECT
+        base64_url,
+        route_id,
+        shape_id,
+        EXTRACT(month FROM service_date) AS month,
+        EXTRACT(year FROM service_date) AS year,
+        COUNT(*) AS n_trips,
+
+    FROM fct_scheduled_trips
+    GROUP BY 1, 2, 3, 4, 5
+    QUALIFY ROW_NUMBER() OVER (PARTITION BY
+                               base64_url, route_id, month, year
+                               ORDER BY n_trips DESC) = 1
+),
+
+fct_monthly_shapes AS (
+
+    SELECT
+        dim_gtfs_datasets.source_record_id,
+        dim_gtfs_datasets.name,
+
+        trips.base64_url,
+        trips.route_id,
+        trips.shape_id,
+        trips.month,
+        trips.year,
+
+        shapes.key AS shape_array_key,
+        shapes.pt_array
+
+    FROM trips_by_shape AS trips
+    INNER JOIN dim_gtfs_datasets
+        ON trips.base64_url = dim_gtfs_datasets.base64_url
+    INNER JOIN dim_shapes_arrays AS shapes
+        ON trips.base64_url = shapes.base64_url
+        AND trips.shape_id = shapes.shape_id
```

Comment:
If you add `month_last_day` as described on line 38, that can be used as the date for this join (can just add another join condition: `AND month_last_day BETWEEN shapes._valid_from AND shapes._valid_to`). Because `dim_shapes_array` is versioned, this join as written would result in fanout; there are multiple rows in `dim_shapes_array` with the same `base64_url` and `shape_id` values (one for each version of that record). Note that if there is a case where the most common shape for the route was deleted and not active at the end of the month, this would be null. Would want to check if/how often that happens.

Reply:
If you first join with `fct_daily_schedule_feeds`, you can use `AND feeds.feed_key = shapes.feed_key` and that will make the join more performant